### PR TITLE
API: tighten DTA/TDA _from_sequence signature

### PR DESCRIPTION
**Pull request by jbrockmendel**
cc @jorisvandenbossche 

need to decide how to handle test_from_2d_ndarray_with_dtype which this currently xfails

### CODE REVIEW

**Review by jreback — CHANGES_REQUESTED** [review]
this looks like a substantial (but good) api change, or is this *just* on master as _from_sequence is actually strict? (in 1.1.x)

**Comment by jbrockmendel** [comment]
> this looks like a substantial (but good) api change, or is this _just_ on master as _from_sequence is actually strict? (in 1.1.x)

not just on master, this makes _from_sequence substantially stricter than 1.1.x

**Comment by jorisvandenbossche** [comment]
Thanks for looking at this!

Now, for the actual behavioural change, `_from_sequence` is still being used for eg `pd.array(.., dtype="datetime64[ns"])` as well, so we first need to decide what behaviour we want for that. 
Otherwise could also keep this in a `_from_sequence_strict` (and rename `_from_sequence_not_strict` to `_from_sequence`)

File: pandas/core/arrays/datetimes.py [hunk 1]
```diff
@@ -300,6 +305,34 @@ def _simple_new(
...
     @classmethod
     def _from_sequence(cls, scalars, dtype=None, copy: bool = False):
+
+        scalars, copy = dtl.ensure_arraylike(scalars, copy)
+
+        if scalars.dtype.kind == "M":
+            pass
+        elif scalars.dtype == object:
+            if isinstance(scalars, ABCMultiIndex):
+                raise TypeError("Cannot create a DatetimeArray from MultiIndex")
```
Comment (jreback) [inline pandas/core/arrays/datetimes.py:315]: this tested?

File: pandas/core/arrays/datetimes.py [hunk 2]
```diff
@@ -300,6 +305,34 @@ def _simple_new(
...
+            # TODO: should go through from_sequence_of_strings?
+            pass
+        elif (
+            is_categorical_dtype(scalars.dtype) and scalars.categories.dtype.kind == "M"
+        ):
+            # TODO: Could also use Categorical[object]
+            #  with inferred_type as above?
+            pass
+        else:
+            msg = f"dtype {scalars.dtype} cannot be converted to datetime64[ns]"
```
Comment (jreback) [inline pandas/core/arrays/datetimes.py:333]: can you ensure we have a test that hits this

File: pandas/core/arrays/timedeltas.py [hunk 3]
```diff
@@ -224,6 +224,22 @@ def _from_sequence(
...
+            inferred = lib.infer_dtype(data)
+            if inferred in ["timedelta64", "timedelta", "empty"]:
+                pass
+            else:
+                raise TypeError(inferred)
+        elif is_string_dtype(data.dtype):
+            # TODO: should go through from_sequence_of_strings?
+            pass
+        else:
+            raise TypeError(data.dtype)
```
Comment (jreback) [inline pandas/core/arrays/timedeltas.py:241]: can you ensure we have a test that hits hits

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jbrockmendel** [comment]
just pushed with a couple new tests that get us to full coverage on the affected methods

File: pandas/tests/frame/test_constructors.py [hunk 4]
```diff
@@ -2869,6 +2869,7 @@ def test_from_tzaware_mixed_object_array(self):
         ]
         assert (res.dtypes == expected_dtypes).all()
 
+    @pytest.mark.xfail(reason="DatetimeArray._from_sequence no longer accepts i8")
```
Comment (jreback) [inline pandas/tests/frame/test_constructors.py:2872]: we need a note for this?

**Review by jreback — CHANGES_REQUESTED** [review]
couple of comments on is this added code tested

File: pandas/tests/frame/test_constructors.py (hunk 4 above)
Comment (jbrockmendel) [reply pandas/tests/frame/test_constructors.py:2872]: the question is if we want to disable/deprecate this or find a way to make it work

**Comment by jreback** [comment]
if u can rebase; i think a couple of questions

**Comment by jbrockmendel** [comment]
cc @WillAyd most recent commit made small edits to json code, can you take a look

File: pandas/tests/frame/test_constructors.py (hunk 4 above)
Comment (jbrockmendel) [reply pandas/tests/frame/test_constructors.py:2901]: is this something we want to support or disallow?

Comment (jorisvandenbossche) [reply pandas/tests/frame/test_constructors.py:2901]: To be clear, the question is not so much about `_from_sequence`, but rather about our public constructors (`Series()`, `Index()`, `array()`).  
Currently, those support converting integers to datetime dtype. For example:

```
pd.Series([1, 2, 3], dtype="datetime64[ns]")
```

works fine (and the same for the other two). 
So the question is if we want to keep this kind of behaviour. Personally I would say there is not much harm in allowing it (also given that eg numpy and pyarrow have the same behaviour), it's only when there are timezones that there could be a bit more ambiguity potentially (but eg for timedelta that's already not relevant).

Comment (jbrockmendel) [reply pandas/tests/frame/test_constructors.py:2901]: I tend to agree.  Any thoughts on where to catch this case?

**Comment by WillAyd** [comment]
lgtm - I think the explicitness here is nice

File: pandas/tests/frame/test_constructors.py (hunk 4 above)
Comment (jorisvandenbossche) [reply pandas/tests/frame/test_constructors.py:2901]: >  Any thoughts on where to catch this case?

Not fully sure what you mean, but (until https://github.com/pandas-dev/pandas/issues/33254 is resolved) doing such conversions is the responsibility of `_from_sequence` (for `pd.array`), so if we want to keep this capability, at the moment it is `_from_sequence` that needs to accept ints.

Comment (jbrockmendel) [reply pandas/tests/frame/test_constructors.py:2901]: > Not fully sure what you mean

I mean that we could make this Series construction work without loosening the _from_sequence restrictions by and instead calling [??] in this case.

> so if we want to keep this capability, at the moment it is _from_sequence that needs to accept ints.

To be clear, is this what you are suggesting we do?  My impression from previous conversations is that you are unambiguously in favor of this tightening.

Comment (jorisvandenbossche) [reply pandas/tests/frame/test_constructors.py]: My earlier suggestion at https://github.com/pandas-dev/pandas/pull/37179#issuecomment-713121153 was to use `_from_sequence`/`_from_sequence_strict` (instead of `_from_sequence_non_strict`/`_from_sequence`), until we decide on better naming for this. So keeping the non-strict behaviour of `_from_sequence` for now. 

> My impression from previous conversations is that you are unambiguously in favor of this tightening.

Yes, for sure. But before we can implement that, we still need to come up with an API for the non-strict conversion as well, as that is still something we need to support. That's the whole discussion of #33254

**Comment by jbrockmendel** [comment]
@jorisvandenbossche any objection to this as a temporary workaround?  im pretty sure It will allow us to clean up a bunch of code in .isin and .equals xref #37528

**Comment by jbrockmendel** [comment]
updated per discussion so that _from_sequence remains non-strict while a new _from_sequence_strict has the hopefully-future behavior of from_sequence.

im hoping to use from_sequence strict to simplify+de-duplicate .equals and .isin

File: pandas/tests/arrays/test_array.py [hunk 5]
```diff
@@ -70,41 +65,42 @@
         (
             pd.DatetimeIndex(["2000", "2001"]),
             np.dtype("datetime64[ns]"),
-            DatetimeArray._from_sequence(["2000", "2001"]),
+            DatetimeArray._from_sequence([pd.Timestamp("2000"), pd.Timestamp("2001")]),
```
Comment (jorisvandenbossche) [inline pandas/tests/arrays/test_array.py:68]: Can you revert the changes to the tests? 
(or otherwise ensure we keep coverage of converting from non-Timestamp scalars)

File: pandas/core/arrays/timedeltas.py [hunk 6]
```diff
@@ -217,6 +217,32 @@ def _simple_new(
...
+        data, copy = dtl.ensure_arraylike(data, copy)
+
+        if data.dtype.kind == "m":
+            pass
+        elif data.dtype == object:
+            inferred = lib.infer_dtype(data)
+            if inferred in ["timedelta64", "timedelta", "empty"]:
+                pass
+            else:
+                raise TypeError(inferred)
```
Comment (jorisvandenbossche) [inline pandas/core/arrays/timedeltas.py:237]: Wondering, should this be a ValueError instead of TypeError? For typical construction we use ValueError (eg `pd.Series(["a"], dtype=float)`), but of course in those cases there is not a single expected type, so ValueError might make more sense there (eg casting from one type to another gives TypeError if it's not a supported cast).

File: pandas/core/arrays/timedeltas.py [hunk 7]
```diff
@@ -217,6 +217,32 @@ def _simple_new(
...
+        elif is_string_dtype(data.dtype):
+            # TODO: should go through from_sequence_of_strings?
+            pass
```
Comment (jorisvandenbossche) [inline pandas/core/arrays/timedeltas.py:240]: Shouldn't we also raise here? (strict means for me also no strings, only actual Timestamp/datetime objects)

**Review by jorisvandenbossche — COMMENTED** [review]
Thanks for the update!

General question: in the tslib conversion functions, those are all "generic" in the sense of accepting a mixture of string, datetime, Timestamp, int, etc (like `array_to_datetime` seems to do) ? We don't have one that is more strict on the input that could be used for this purpose? (instead of checking up front with infer_dtype)

File: pandas/tests/arrays/test_array.py (hunk 5 above)
Comment (jbrockmendel) [reply pandas/tests/arrays/test_array.py:68]: i was on the fence about this, glad you tipped the scales.  will update.

File: pandas/core/arrays/timedeltas.py (hunk 6 above)
Comment (jbrockmendel) [reply pandas/core/arrays/timedeltas.py:237]: sure

File: pandas/tests/arrays/test_array.py (hunk 5 above)
Comment (jbrockmendel) [reply pandas/tests/arrays/test_array.py:68]: i was on the fence about this, glad you tipped the scales.  will update.

File: pandas/core/arrays/timedeltas.py (hunk 7 above)
Comment (jbrockmendel) [reply pandas/core/arrays/timedeltas.py]: yep, updated

**Comment by jbrockmendel** [comment]
> We don't have one that is more strict on the input that could be used for this purpose?

Good idea, we can use `tslibs.conversion.datetime_to_datetime64`.  AFAICT we dont have an equivalent for timedelta, but it wouldn't be hard to implement one.  `libperiod.extract_ordinals` can also be adapted to do the same thing for PeriodArray (separate PR).

Will update.

**Comment by jbrockmendel** [comment]
mothballing to clear the queue while i address comments locally
